File: python/backtracking/combination_sum.py
from typing import List


class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []

        def backtrack(start: int, path: List[int], remaining: int):
            if remaining == 0:
                result.append(list(path))

                return


            for i in range(start,  len(candidates)):
                if candidates[i] >  remaining:
                    continue


                path.append(candidates[i])

                backtrack(i, path, remaining-candidates[i])

                path.pop()


        backtrack(0, [], target)


        return result

File: golang/backtracking/combination_sum.go
package main

func combinationSum(candidates []int, target int) [][]int {
	var result [][]int

	var backtrack func(start int, path []int, remaining int)

	backtrack = func(start int, path []int, remaining int) {
		if remaining == 0 {
			combination := make([]int, len(path))

			copy(combination, path)
			result = append(result, combination)

			return
		}

		for i := start; i < len(candidates); i++ {
			if candidates[i] > remaining {
				continue
			}

			path = append(path, candidates[i])

			backtrack(i, path, remaining-candidates[i])

			path = path[:len(path)-1]
		}
	}

	backtrack(0, []int{}, target)

	return result

}
